sql/src/test/java/io/crate/operation/fetch/PositionalBucketMergerTest.java

```diff
@@ -0,0 +1,119 @@
+import io.crate.core.collections.Bucket;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.RowN;
+import io.crate.operation.projectors.CollectingProjector;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.is;
+
+public class PositionalBucketMergerTest extends RandomizedTest {
+
+    @Test
+    public void testConcurrentSetNextBucket() throws Exception {
+        /**
+         * Propagate 2 buckets of the same upstream concurrent, every bucket is ordered.
+         */
+        int numUpstreams = 2;
+
+        CollectingProjector resultProvider = new CollectingProjector();
+        final PositionalBucketMerger bucketMerger = new PositionalBucketMerger(resultProvider, numUpstreams, 1);
+
+        final List<List<List<Object[]>>> bucketsPerUpstream = new ArrayList<>(numUpstreams);
+        List<List<Object[]>> upstream1 = new ArrayList<>(2);
+        upstream1.add(ImmutableList.of(new Object[]{4L}, new Object[]{6L}, new Object[]{7L}));
+        upstream1.add(ImmutableList.of(new Object[]{0L}, new Object[]{1L}, new Object[]{3L}));
+        bucketsPerUpstream.add(upstream1);
+
+        List<List<Object[]>> upstream2 = new ArrayList<>(2);
+        upstream2.add(ImmutableList.of(new Object[]{2L}, new Object[]{5L}));
+        upstream2.add(ImmutableList.of(new Object[]{8L}, new Object[]{9L}));
+        bucketsPerUpstream.add(upstream2);
+
+        final CountDownLatch latch = new CountDownLatch(numUpstreams*2);
+        final ExecutorService executorService = Executors.newScheduledThreadPool(numUpstreams);
+        for (int i = 0; i < bucketsPerUpstream.size(); i++) {
+            final int upstreamId = i;
+
+            for (final List<Object[]> bucket : bucketsPerUpstream.get(i)) {
+                bucketMerger.registerUpstream(null);
+                executorService.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        List<Row> rows1 = new ArrayList<>();
+                        for (Object[] row : bucket) {
+                            rows1.add(new PositionalRowDelegate(new RowN(row), (long) row[0]));
+                        }
+                        try {
+                            bucketMerger.setNextBucket(rows1, upstreamId);
+                        } catch (Exception e) {
+                            fail(e.getMessage());
+                        }
+                        bucketMerger.finish();
+                        latch.countDown();
+                    }
+                });
+            }
+        }
+        latch.await();
+        executorService.shutdown();
+        bucketMerger.finish();
+
+        final CountDownLatch latch1 = new CountDownLatch(1);
```

Comment:
can you use a SettableFuture here, so that a possible error can be propagated from another thread (possibly) and fail in the test thread?